osutil/epoll/epoll.go

```diff
@@ -0,0 +1,192 @@
+	countMutex        sync.Mutex
+}
+
+// Open opens an event monitoring descriptor.
+func Open() (*Epoll, error) {
+	fd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %w", err)
+	}
+	e := &Epoll{
+		fd:                fd,
+		registeredFdCount: 0,
+	}
+	runtime.SetFinalizer(e, func(e *Epoll) {
+		if e.fd != -1 {
+			e.Close()
+		}
+	})
+	return e, nil
+}
+
+// Close closes the event monitoring descriptor.
+func (e *Epoll) Close() error {
+	runtime.SetFinalizer(e, nil)
+	e.countMutex.Lock()
+	fd := e.fd
+	e.fd = -1
+	e.registeredFdCount = 0
+	e.countMutex.Unlock()
+	return unix.Close(fd)
+}
+
+// Fd returns the integer unix file descriptor referencing the open file.
+func (e *Epoll) Fd() int {
+	return e.fd
+}
+
+// RegisteredFdCount returns the number of file descriptors which are currently
+// registered to the epoll instance.
+func (e *Epoll) RegisteredFdCount() int {
+	return e.registeredFdCount
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount += 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Deregister removes the given file descriptor from the epoll instance.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_DEL for details.
+func (e *Epoll) Deregister(fd int) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_DEL, fd, &unix.EpollEvent{})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount -= 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Modify changes the set of monitored I/O readiness events of a previously registered file descriptor.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_MOD for details.
+func (e *Epoll) Modify(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_MOD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Event describes an IO readiness event on a specific file descriptor.
+type Event struct {
+	Fd        int
+	Readiness Readiness
+}
+
+// WaitTimeout blocks and waits with the given timeout for arrival of events on any of the added file descriptors.
+//
+// A msec value of -1 disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeout(duration time.Duration) ([]Event, error) {
+	msec := int(duration.Milliseconds())
+	if duration < 0 {
+		msec = -1
+	}
+	n := 0
+	var err error
+	var sysEvents []unix.EpollEvent
+	for {
+		sysEvents = make([]unix.EpollEvent, e.registeredFdCount)
```

Comment:
this is not locking, in theory something could register a new fd while the wait is running? we need to take the lock or simplify things and expect Epoll to be used by a single goroutine, I'm not sure what usage we need

Reply:
mmh, given how epoll_wait works maybe this is not a big problem, but is still the case that we need to read registeredFdCount with the lock. Given that is only a counter though we usually use the functions from https://pkg.go.dev/sync/atomic instead of a mutex

Reply:
Yes, I wrote about this in a commit message: https://github.com/snapcore/snapd/pull/12963/commits/21582cd78d1b35779c728e49cf4c8d8ec2fb5417

In short, we don't mind if another fd is registered while the epoll_wait syscall is still waiting. If there are multiple events waiting before the wait call, then there will be enough entries in the buffer to save them. Otherwise, the first single event to occur should cause the epoll_wait to return, and we should only need one entry in the buffer anyway.

Worst case, if there's an event waiting for every registered event and we manage to register a new fd and get activity on it between the allocation of the event buffer and the call to epoll_wait, the syscall will fill the buffer with events, and on the next epoll_wait call, the event from the newly-registered fd will be returned. There is no requirement that epoll_wait return every pending event, just at most n, where n is the number of entries in the event buffer passed into the syscall. The `x/sys/unix.EpollWait()` function abstracts this away by setting `n = len(sysEvents)`, which is the buffer which we pass into the syscall.

I'm totally open to locking for reads, but why do we need to read with a lock in this case? If `count = x` and there is a race condition with `Wait()` and `Register()`, the count will either be `x` or `x+1`, either of which is acceptable. If we do lock the read, then between unlocking and calling the epoll_wait syscall, `Register()` could be called, incrementing the count anyway, so I don't really know how locking would help. It wouldn't hurt to lock for reads, but then I would likely want to use a RWLock instead, and would generally add a bit of complexity.

I would have liked to use `sync/atomic/Int32`, but it requires go 1.19.  I considered using `sync/atomic/AddInt32()`, but didn't love storing the value as `int32` without any indication besides other uses of `AddInt32()` that it should be protected. Happy to use this instead though and remove the mutex.

FWIW, golang itself just passes in a buffer of size 128 or 256 or something arbitrarily to its internal epoll_wait calls. Previously, Zygmunt just passed in a buffer of 10 entries always. Tying the buffer size to the number of registered FDs is more a matter of convenience/slight memory efficiency.

Reply:
I replaced the mutex with atomic operations for add, store, and load of the int32. Thanks for the suggestions!